pkg/services/ngalert/notifier/channels/email_test.go

```diff
@@ -106,3 +108,193 @@ func TestEmailNotifier(t *testing.T) {
+			alerts: []*types.Alert{
+				{
+					Alert: model.Alert{
+						Labels:      model.LabelSet{"alertname": "FiringOne", "severity": "warning"},
+						Annotations: model.LabelSet{"runbook_url": "http://fix.me", "__dashboardUid__": "abc", "__panelId__": "5"},
+					},
+				},
+				{
+					Alert: model.Alert{
+						Labels:      model.LabelSet{"alertname": "FiringTwo", "severity": "critical"},
+						Annotations: model.LabelSet{"runbook_url": "http://fix.me", "__dashboardUid__": "abc", "__panelId__": "5"},
+					},
+				},
+			},
+			messageTmpl: "",
+			expSubject:  "[FIRING:2]  ",
+			expSnippets: []string{
+				"Firing: 2 alerts",
+				"<li>alertname: FiringOne</li><li>severity: warning</li>",
+				"<li>alertname: FiringTwo</li><li>severity: critical</li>",
+				"<a href=\"http://fix.me\"",
+				"<a href=\"http://localhost/base/d/abc",
+				"<a href=\"http://localhost/base/d/abc?viewPanel=5",
+			},
+		},
+		{
+			name: "message containing HTML gets HTMLencoded",
+			alerts: []*types.Alert{
+				{
+					Alert: model.Alert{
+						Labels:      model.LabelSet{"alertname": "AlwaysFiring", "severity": "warning"},
+						Annotations: model.LabelSet{"runbook_url": "http://fix.me", "__dashboardUid__": "abc", "__panelId__": "5"},
+					},
+				},
+			},
+			messageTmpl: `<marquee>Hi, this is a custom template.</marquee>
+				{{ if gt (len .Alerts.Firing) 0 }}
+					<ol>
+					{{range .Alerts.Firing }}<li>Firing: {{ .Labels.alertname }} at {{ .Labels.severity }} </li> {{ end }}
+					</ol>
+				{{ end }}`,
+			expSubject: "[FIRING:1]  (AlwaysFiring warning)",
+			expSnippets: []string{
+				"&lt;marquee&gt;Hi, this is a custom template.&lt;/marquee&gt;",
+				"&lt;li&gt;Firing: AlwaysFiring at warning &lt;/li&gt;",
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			emailNotifier := createSut(t, c.messageTmpl, emailTmpl, bus)
+
+			ok, err := emailNotifier.Notify(context.Background(), c.alerts...)
+			require.NoError(t, err)
+			require.True(t, ok)
+
+			sentMsg := ns.MailQueuePop()
+
+			require.NotNil(t, sentMsg)
+
+			require.Equal(t, "\"Grafana Admin\" <from@address.com>", sentMsg.From)
+			require.Equal(t, sentMsg.To[0], "someops@example.com")
+
+			require.Equal(t, c.expSubject, sentMsg.Subject)
+
+			require.Contains(t, sentMsg.Body, "text/html")
+			html := sentMsg.Body["text/html"]
+			require.NotNil(t, html)
+
+			for _, s := range c.expSnippets {
+				require.Contains(t, html, s)
+			}
+		})
+	}
+}
+
+func createCoreEmailService(t *testing.T) (*notifications.NotificationService, *bus.InProcBus) {
+	t.Helper()
+
+	setting.StaticRootPath = "../../../public/"
+	setting.BuildVersion = "4.0.0"
+
+	ns := &notifications.NotificationService{}
+	bus := bus.New()
+	cfg := setting.NewCfg()
+	cfg.Smtp.Enabled = true
+	cfg.Smtp.TemplatesPatterns = []string{"/home/alexweav/git/grafana/public/emails/*.html", "/home/alexweav/git/grafana/public/emails/*.txt"}
```

Comment:
Ha! Good catch, I intended to clean this up before submission but forgot to add a TODO 😀

Reply:
Fixed. This took a sec to figure out (which is why I hacked in an absolute path to begin with) the right relpath ended up being `../../../../../public/emails/*.html`. This is because:

- The templates themselves are templated - the ones in emails/blah end up getting preprocessed prior to install, and the result gets codegen'd and committed under `public/emails`.
- `os.Getwd()` from inside a Go test returns the directory containing the test file. So, each test acts as if it's being executed out of the folder where it's defined. (this is different from other languages, but I kind of like it :grin:)